From b7998694d74171925cddeedf34ac005bf90b136f Mon Sep 17 00:00:00 2001
From: Sam Leffler <sleffler@chromium.org>
Date: Thu, 31 Mar 2011 17:39:06 -0700
Subject: [PATCH] CHROMIUMOS: wpa_supplicant: Accept raw PMK in new DBus API

"psk" field, in set_network_properties(), was always getting quoted even when
containing a raw key (64 characters length representing the hexadecimal value
of the raw key).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@nokia.com>

(pulled from upstream patch)

BUG=chromium-os:13727
TEST=WiFiManager/022CheckWPA_RawPMK

Review URL: http://codereview.chromium.org/6783033

Change-Id: I1f9f5215b02a8cc0e532ca105f68a8828736ce21
---
 wpa_supplicant/dbus/dbus_new_handlers.c | 24 +++++++++++++++++++-----
 1 file changed, 19 insertions(+), 5 deletions(-)

diff --git a/wpa_supplicant/dbus/dbus_new_handlers.c b/wpa_supplicant/dbus/dbus_new_handlers.c
index 9f6c4a3..12ac0e2 100644
--- a/wpa_supplicant/dbus/dbus_new_handlers.c
+++ b/wpa_supplicant/dbus/dbus_new_handlers.c
@@ -144,7 +144,9 @@ static const char *dont_quote[] = {
 	"bssid", "scan_freq", "freq_list", NULL
 };
 
-static dbus_bool_t should_quote_opt(const char *key)
+static dbus_bool_t should_quote_opt(const char *key,
+				    size_t value_length,
+				    char *value)
 {
 	int i = 0;
 	while (dont_quote[i] != NULL) {
@@ -152,6 +154,17 @@ static dbus_bool_t should_quote_opt(const char *key)
 			return FALSE;
 		i++;
 	}
+
+	/*
+	 * Do not quote psk value which is a raw key.
+	 */
+	if ((os_strcmp(key, "psk") == 0) && (value_length == PMK_LEN*2)) {
+		u8 tmp_buf[PMK_LEN];
+
+		if (hexstr2bin(value, tmp_buf, PMK_LEN) == 0)
+			return FALSE;
+	}
+
 	return TRUE;
 }
 
@@ -221,11 +234,12 @@ dbus_bool_t set_network_properties(struct wpa_supplicant *wpa_s,
 			if (ret <= 0)
 				goto error;
 		} else if (entry.type == DBUS_TYPE_STRING) {
-			if (should_quote_opt(entry.key)) {
-				size = os_strlen(entry.str_value);
-				if (size <= 0)
-					goto error;
+			size = os_strlen(entry.str_value);
+			if (size <= 0)
+				goto error;
 
+			if (should_quote_opt(entry.key, size,
+					     entry.str_value)) {
 				size += 3;
 				value = os_zalloc(size);
 				if (value == NULL)
-- 
2.2.0.rc0.207.ga3a616c

